tools/reorder_docids.hpp

```diff
@@ -0,0 +1,319 @@
+    gsl::span<std::uint32_t const> mapping)
+
+{
+    auto doc_buffer = Payload_Vector_Buffer::from_file(input_lexicon);
+    auto documents = Payload_Vector<std::string>(doc_buffer);
+    std::vector<std::string> reordered_documents(documents.size());
+    pisa::progress doc_reorder("Reordering documents vector", documents.size());
+    for (size_t i = 0; i < documents.size(); ++i) {
+        reordered_documents[mapping[i]] = documents[i];
+        doc_reorder.update(1);
+    }
+    encode_payload_vector(reordered_documents.begin(), reordered_documents.end()).to_file(output_lexicon);
+}
+
+auto reorder_sizes(
+    binary_collection const& input_sizes,
+    std::uint64_t num_docs,
+    gsl::span<std::uint32_t const> mapping,
+    std::string_view output_basename)
+{
+    pisa::progress progress("Reordering document sizes", num_docs);
+    auto sizes = *input_sizes.begin();
+    if (sizes.size() != num_docs) {
+        throw std::invalid_argument("Invalid sizes file");
+    }
+
+    auto size_sequence = gsl::span(sizes.begin(), sizes.size());
+    std::vector<std::uint32_t> new_sizes(num_docs);
+    for (size_t i = 0; i < num_docs; ++i) {
+        new_sizes[mapping[i]] = size_sequence[i];
+        progress.update(1);
+    }
+
+    std::ofstream output_sizes(fmt::format("{}.sizes", output_basename));
+    emit(output_sizes, new_sizes.size());
+    emit(output_sizes, new_sizes.data(), num_docs);
+}
+
+void reorder_from_mapping(
+    binary_freq_collection const& input_collection,
+    binary_collection const& input_sizes,
+    std::string output_basename,
+    std::optional<std::string> const& document_lexicon,
+    std::optional<std::string> const& reordered_document_lexicon,
+    gsl::span<std::uint32_t const> mapping)
+{
+    auto num_docs = input_collection.num_docs();
+    reorder_sizes(input_sizes, num_docs, mapping, output_basename);
+    reorder_postings(input_collection, output_basename, mapping);
+    if (document_lexicon) {
+        reorder_lexicon(*document_lexicon, *reordered_document_lexicon, mapping);
+    }
+}
+
+auto reorder_random(
+    std::string const& input_basename,
+    std::string output_basename,
+    std::optional<std::string> const& document_lexicon,
+    std::optional<std::string> const& reordered_document_lexicon,
+    std::uint64_t seed) -> int
+{
+    spdlog::info("Computing random permutation");
+    binary_freq_collection input_collection(input_basename.c_str());
+    auto num_docs = input_collection.num_docs();
+    std::vector<std::uint32_t> mapping(num_docs);
+    std::iota(mapping.begin(), mapping.end(), uint32_t());
+    std::shuffle(mapping.begin(), mapping.end(), std::mt19937{seed});
+
+    binary_collection input_sizes(fmt::format("{}.sizes", input_basename).c_str());
+    reorder_from_mapping(
+        input_collection,
+        input_sizes,
+        output_basename,
+        document_lexicon,
+        reordered_document_lexicon,
+        mapping);
+    return 0;
+}
+
+auto reorder_url(
```

Comment:
This is almost more generic than URL sorting right? You could pass a file containing a string for each document, where the string is the target sort key, and this would still work. Correct? Perhaps we should modify it to be more generic than URL?

Reply:
Of course, users could just do this with the pre-defined mapping component as well.

Reply:
True, this one and the one with defined mapping is very similar, only this is really a reverse mapping really (or the other one is, depending on the way you look at it). So really, these two could co-exist but maybe it would be confusing which is which? Not sure. I wanted to have an option "by URL" so that it's easier to think about, but the problem currently is that you still have to pass that url file.

Bottom line, it's not a bad idea to make it as generic as it can be. If there's an easier way to just do url without passing a file, then we can do it then. For now, though, I we would need to think hard about naming these options to make it clear what it is. Maybe `--with-mapping` vs. `--by-label`. Something descriptive. Any suggestions?